### PR TITLE
Update `smallvec` to 1.8.1.

### PR DESCRIPTION
**Pull request by nnethercote**
This pulls in https://github.com/servo/rust-smallvec/pull/282, which
gives some small wins for rustc.

r? @lqd

### CODE REVIEW

**Comment by nnethercote** [comment]
I updated all `smallvec` occurrences in `Cargo.toml` files. The resulting `Cargo.lock` files are as follows:
```
[gulf:~/dev/rust1] grep -r -s -A1 --include=Cargo.lock "name = .*smallvec"
Cargo.lock:name = "smallvec"
Cargo.lock-version = "1.8.1"
--
compiler/rustc_codegen_cranelift/Cargo.lock:name = "smallvec"
compiler/rustc_codegen_cranelift/Cargo.lock-version = "1.8.1"
--
src/tools/miri/Cargo.lock:name = "smallvec"
src/tools/miri/Cargo.lock-version = "1.7.0"
--
src/tools/rust-analyzer/Cargo.lock:name = "smallvec"
src/tools/rust-analyzer/Cargo.lock-version = "1.8.0"
--
src/tools/rls/Cargo.lock:name = "smallvec"
src/tools/rls/Cargo.lock-version = "1.7.0"
```
I.e. the main `Cargo.lock` and the one for cranelift were updated, but the miri, rust-analyzer and rls ones were not. I'm not sure if that's expected. I also don't fully understand how `src/tools/rustc-workspace-hack/Cargo.toml` interacts with those. I'm happy to hear suggestions on anything I should or should not do relating to those tools.

**Comment by nnethercote** [comment]
Perf results are as expected: a win for `tt-muncher` and doesn't affect anything else.

**Comment by eddyb** [comment]
> I.e. the main `Cargo.lock` and the one for cranelift were updated, but the miri, rust-analyzer and rls ones were not.

miri and the rest are presumably all git submodules, so their `Cargo.lock` files are not in this repo.
If you locally change any files in there, then run `x.py` commands, they will likely get reverted to the last commit in that specific submodule.

File: compiler/rustc_apfloat/Cargo.toml [hunk 1]
```diff
@@ -5,4 +5,4 @@ edition = "2021"
 
 [dependencies]
 bitflags = "1.2.1"
-smallvec = { version = "1.6.1", features = ["union", "may_dangle"] }
+smallvec = { version = "1.8.1", features = ["union", "may_dangle"] }
```
Comment (eddyb) [inline compiler/rustc_apfloat/Cargo.toml:8]: Not a source change, checking off on `rustc_apfloat` :heavy_check_mark:.

**Pull request merged**